source/common/router/router.cc

```diff
@@ -679,84 +748,132 @@ Filter::streamResetReasonToResponseFlag(Http::StreamResetReason reset_reason) {
+  UpstreamRequest* upstream_request_tmp;
+  for (auto it = upstream_requests_.cbegin(); it != upstream_requests_.cend(); it++) {
+    upstream_request_tmp = it->get();
+    if (upstream_request_tmp != upstream_request) {
+      if (!upstream_request_tmp->encode_complete_ || !upstream_request_tmp->decode_complete_) {
+        upstream_request_tmp->resetStream();
+        if (upstream_request_tmp->upstream_host_) {
+          upstream_request_tmp->upstream_host_->stats().rq_hedge_abandoned_.inc();
+        }
+        cluster_->stats().upstream_rq_hedge_abandoned_.inc();
+      }
+    }
+  }
 }
 
 void Filter::onUpstreamHeaders(uint64_t response_code, Http::HeaderMapPtr&& headers,
-                               bool end_stream) {
+                               UpstreamRequest* upstream_request, bool end_stream) {
   ENVOY_STREAM_LOG(debug, "upstream headers complete: end_stream={}", *callbacks_, end_stream);
 
-  upstream_request_->upstream_host_->outlierDetector().putHttpResponseCode(response_code);
+  upstream_request->upstream_host_->outlierDetector().putHttpResponseCode(response_code);
 
   if (headers->EnvoyImmediateHealthCheckFail() != nullptr) {
-    upstream_request_->upstream_host_->healthChecker().setUnhealthy();
+    upstream_request->upstream_host_->healthChecker().setUnhealthy();
   }
 
+  bool could_not_retry = false;
+
+  // Check if this upstream request was already retried, for instance after
+  // hitting a per try timeout. Don't retry it if we already have.
   if (retry_state_) {
-    // Notify retry modifiers about the attempted host.
-    retry_state_->onHostAttempted(upstream_request_->upstream_host_);
-
-    const RetryStatus retry_status =
-        retry_state_->shouldRetryHeaders(*headers, [this]() -> void { doRetry(); });
-    // Capture upstream_host since setupRetry() in the following line will clear
-    // upstream_request_.
-    const auto upstream_host = upstream_request_->upstream_host_;
-    if (retry_status == RetryStatus::Yes && setupRetry(end_stream)) {
-      Http::CodeStats& code_stats = httpContext().codeStats();
-      code_stats.chargeBasicResponseStat(cluster_->statsScope(), "retry.",
-                                         static_cast<Http::Code>(response_code));
-      upstream_host->stats().rq_error_.inc();
-      return;
-    } else if (retry_status == RetryStatus::NoOverflow) {
-      callbacks_->streamInfo().setResponseFlag(StreamInfo::ResponseFlag::UpstreamOverflow);
-    } else if (retry_status == RetryStatus::NoRetryLimitExceeded) {
-      callbacks_->streamInfo().setResponseFlag(
-          StreamInfo::ResponseFlag::UpstreamRetryLimitExceeded);
+    if (upstream_request->retried_) {
+      // We already retried this request (presumably for a per try timeout) so
+      // we definitely won't retry it again. Check if we would have retried it
+      // if we could.
+      could_not_retry = retry_state_->wouldRetryFromHeaders(*headers);
```

Comment:
If this variable says that we wanted to retry but couldn't, why is it set to true when we hit the retry count limit?

Reply:
bump

Reply:
it means like "request was bad but won't/can't retry it". That could be either because we already retried it w/ hedging or because we hit the limit